core/file/flock_spec.rb

```diff
@@ -30,33 +30,48 @@
   it "returns false if trying to lock an exclusively locked file" do
     @file.flock File::LOCK_EX
 
-    File.open(@name, "w") do |f2|
-      f2.flock(File::LOCK_EX | File::LOCK_NB).should == false
-    end
+    ruby_exe(<<-END_OF_CODE, escape: true).should == "false"
+      File.open('#{@name}', "w") do |f2|
+        print f2.flock(File::LOCK_EX | File::LOCK_NB).to_s
+      end
+    END_OF_CODE
   end
 
   it "blocks if trying to lock an exclusively locked file" do
     @file.flock File::LOCK_EX
 
-    running = false
     t = Thread.new do
-      ScratchPad << :before
+      ruby_exe(<<-END_OF_CODE, escape: true)
+        File.open('#{@name}', "w") do |f2|
+          begin
+            f2.puts Process.pid
+            f2.puts "before"
+            f2.flush
+            f2.flock(File::LOCK_EX)
+            f2.puts "after"
+          rescue Interrupt
+            f2.puts "interrupt"
+          end
+        end
+      END_OF_CODE
+    end
 
-      running = true
-      File.open(@name, "w") do |f2|
-        f2.flock(File::LOCK_EX)
+    begin
+      @file.seek(0, IO::SEEK_SET)
+      begin
+        s = @file.read_nonblock(127)
+      rescue EOFError, Errno::EAGAIN, Errno::EWOULDBLOCK
       end
+    end until /before/ =~ s.to_s
```

Comment:
> Is it possible that the read_nonblock returns only a part of before (like bef)?
Yes, possible (though rare). This may also happen to normal IO methods (e.g. gets).

> Should we accumulate the results in a buffer then?
It is possible to do so, but this would make code complicated.
Moreover, in my experience, when a file is modified from other processes, I think it is safe to set the file position again.

>Or maybe using two calls to IO#gets would be simpler?
This does not support corner case to read the file when writing to the file is partially done.
Writing to the file and reading from the file is asynchronous and any half-baked state could occur.